From: Jan Horak <jhorak@redhat.com>
Date: Tue, 17 Dec 2024 17:18:33 +0100
Subject: webextensions: add a portal for managing WebExtensions native
 messaging servers.

Rebase, fix and continue work on webextensions: add a portal for managing WebExtensions
native messaging servers:
https://github.com/flatpak/xdg-desktop-portal/pull/705

This commit builds on the work done in the original MR authored by @jhenstridge
but resolves pending items and brings it closer to completion.

This is intended to provide a way for a confined web browser to start
native code helpers for their extensions. At present it can start the
servers installed on the host system. But in future this could be
extended to cover sandboxed native messaging servers too.

Fixes: #769

Origin: https://github.com/swick/xdg-desktop-portal/tree/wip/webextensions
---
 data/meson.build                              |   1 +
 data/org.freedesktop.portal.WebExtensions.xml | 158 +++++
 meson.build                                   |   1 +
 po/POTFILES.in                                |   1 +
 src/meson.build                               |   1 +
 src/web-extensions.c                          | 875 ++++++++++++++++++++++++++
 src/web-extensions.h                          |  24 +
 src/xdg-desktop-portal.c                      |   4 +
 src/xdp-request.c                             |  12 +
 tests/conftest.py                             |  31 +-
 tests/meson.build                             |   1 +
 tests/test_webextensions.py                   | 147 +++++
 12 files changed, 1248 insertions(+), 8 deletions(-)
 create mode 100644 data/org.freedesktop.portal.WebExtensions.xml
 create mode 100644 src/web-extensions.c
 create mode 100644 src/web-extensions.h
 create mode 100644 tests/test_webextensions.py

diff --git a/data/meson.build b/data/meson.build
index 9f2eaf1..174341c 100644
--- a/data/meson.build
+++ b/data/meson.build
@@ -37,6 +37,7 @@ portal_sources = files(
   'org.freedesktop.portal.Trash.xml',
   'org.freedesktop.portal.Usb.xml',
   'org.freedesktop.portal.Wallpaper.xml',
+  'org.freedesktop.portal.WebExtensions.xml',
 )
 
 portal_host_sources = files(
diff --git a/data/org.freedesktop.portal.WebExtensions.xml b/data/org.freedesktop.portal.WebExtensions.xml
new file mode 100644
index 0000000..2725c02
--- /dev/null
+++ b/data/org.freedesktop.portal.WebExtensions.xml
@@ -0,0 +1,158 @@
+<?xml version="1.0"?>
+<!--
+ Copyright (C) 2022 Canonical Ltd
+
+ This library is free software; you can redistribute it and/or
+ modify it under the terms of the GNU Lesser General Public
+ License as published by the Free Software Foundation; either
+ version 2 of the License, or (at your option) any later version.
+
+ This library is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ Lesser General Public License for more details.
+
+ You should have received a copy of the GNU Lesser General Public
+ License along with this library. If not, see <http://www.gnu.org/licenses/>.
+-->
+
+<node name="/" xmlns:doc="http://www.freedesktop.org/dbus/1.0/doc.dtd">
+  <!--
+      org.freedesktop.portal.WebExtensions:
+      @short_description: WebExtensions portal
+
+      The WebExtensions portal allows sandboxed web browsers to start
+      native messaging hosts installed on the host system.
+
+      Accompanying documentation for Firefox's implementation is
+      available: `Native messaging for a strictly-confined Firefox
+      <https://firefox-source-docs.mozilla.org/toolkit/components/extensions/webextensions/native-messaging-portal-design.html>`_.
+
+      This documentation describes version 1 of this interface.
+  -->
+  <interface name="org.freedesktop.portal.WebExtensions">
+    <!--
+        CreateSession:
+        @options: Vardict with optional further information
+        @session_handle: Object path for the #org.freedesktop.portal.Session created by this call.
+
+        Create a web extensions session. A successfully created
+        session can at any time be closed using
+        org.freedesktop.portal.Session::Close, or may at any time be
+        closed by the portal implementation, which will be signalled
+        via org.freedesktop.portal.Session::Closed.
+
+        To close a session, the browser should:
+
+        1. close the process's stdin/stdout/stderr file descriptors
+           obtained from the portal;
+        2. wait for a D-Bus Closed signal from the portal on the
+           org.freedesktop.portal.Session object (which will be
+           triggered on SIGCHLD via the g_child_watch_add_full
+           handler); and
+        3. if the Closed signal from the portal doesn't come in time,
+           call the Close method on the org.freedesktop.portal.Session
+           object.
+
+        Supported keys in the @options vardict include:
+        <variablelist>
+          <varlistentry>
+            <term>mode s</term>
+            <listitem><para>
+              A string indicating which behaviour the portal should
+              use when locating and starting native messaging
+              hosts. Valid values are "mozilla" and "chromium". By
+              default, mozilla behaviour is used.
+            </para></listitem>
+          </varlistentry>
+          <varlistentry>
+            <term>session_handle_token s</term>
+            <listitem><para>
+              A string that will be used as the last element of the session handle. Must be a valid
+              object path element. See the #org.freedesktop.portal.Session documentation for
+              more information about the session handle.
+            </para></listitem>
+          </varlistentry>
+        </variablelist>
+    -->
+    <method name="CreateSession">
+      <annotation name="org.qtproject.QtDBus.QtTypeName.In0" value="QVariantMap"/>
+      <arg type="a{sv}" name="options" direction="in"/>
+      <arg type="o" name="session_handle" direction="out"/>
+    </method>
+    <!--
+        GetManifest:
+        @session_handle: Object path for the #org.freedesktop.portal.Session object
+        @name: name of the native messaging host
+        @extension_or_origin: extension ID or origin URI identifying the extension
+        @json_manifest: the JSON manifest for the native messaging host
+
+        Return the JSON manifest of the native messaging host that
+        Start would invoke.
+    -->
+    <method name="GetManifest">
+      <arg type="o" name="session_handle" direction="in"/>
+      <arg type="s" name="name" direction="in"/>
+      <arg type="s" name="extension_or_origin" direction="in"/>
+      <arg type="s" name="json_manifest" direction="out"/>
+    </method>
+    <!--
+        Start:
+        @session_handle: Object path for the #org.freedesktop.portal.Session object
+        @name: name of the native messaging host
+        @extension_or_origin: extension ID or origin URI identifying the extension
+        @options: Vardict with optional further information
+        @handle: Object path for the #org.freedesktop.portal.Request object representing this call
+
+        Start the named native messaging host. The caller must
+        indicate the requesting web extension (either by extension ID
+        for Firefox, or origin URI for Chrome), which will be matched
+        against the host's access control list.
+
+        If the host can't be started, or invalid data is provided,
+        the session will be closed.
+
+        Supported keys in the @options vardict include:
+        <variablelist>
+          <varlistentry>
+            <term>handle_token s</term>
+            <listitem><para>
+              A string that will be used as the last element of the @handle. Must be a valid
+              object path element. See the #org.freedesktop.portal.Request documentation for
+              more information about the @handle.
+            </para></listitem>
+          </varlistentry>
+        </variablelist>
+    -->
+    <method name="Start">
+      <arg type="o" name="session_handle" direction="in"/>
+      <arg type="s" name="name" direction="in"/>
+      <arg type="s" name="extension_or_origin" direction="in"/>
+      <annotation name="org.qtproject.QtDBus.QtTypeName.In3" value="QVariantMap"/>
+      <arg type="a{sv}" name="options" direction="in"/>
+      <arg type="o" name="handle" direction="out"/>
+    </method>
+    <!--
+        GetPipes:
+        @session_handle: Object path for the #org.freedesktop.portal.Session object
+        @options: Vardict with optional further information
+        @stdin: File descriptor representing the hosts's stdin.
+        @stdout: File descriptor representing the host's stdout.
+        @stderr: File descriptor representing the host's stderr.
+
+        Retrieve file descriptors for the native messaging host
+        identified by the session. This method should only be called
+        after the Start request recveives a successful response.
+    -->
+    <method name="GetPipes">
+      <annotation name="org.gtk.GDBus.C.UnixFD" value="true"/>
+      <arg type="o" name="session_handle" direction="in"/>
+      <annotation name="org.qtproject.QtDBus.QtTypeName.In1" value="QVariantMap"/>
+      <arg type="a{sv}" name="options" direction="in"/>
+      <arg type="h" name="stdin" direction="out"/>
+      <arg type="h" name="stdout" direction="out"/>
+      <arg type="h" name="stderr" direction="out"/>
+    </method>
+    <property name="version" type="u" access="read"/>
+  </interface>
+</node>
diff --git a/meson.build b/meson.build
index fa6b7ff..cb72f2b 100644
--- a/meson.build
+++ b/meson.build
@@ -82,6 +82,7 @@ config_h = configuration_data()
 config_h.set('_GNU_SOURCE', 1)
 config_h.set_quoted('G_LOG_DOMAIN', 'xdg-desktop-portal')
 config_h.set_quoted('DATADIR', datadir)
+config_h.set_quoted('LIBDIR', libdir)
 config_h.set_quoted('LIBEXECDIR', libexecdir)
 config_h.set_quoted('LOCALEDIR', localedir)
 config_h.set_quoted('SYSCONFDIR', sysconfdir)
diff --git a/po/POTFILES.in b/po/POTFILES.in
index d514b21..fd32325 100644
--- a/po/POTFILES.in
+++ b/po/POTFILES.in
@@ -7,3 +7,4 @@ src/screenshot.c
 src/settings.c
 src/usb.c
 src/wallpaper.c
+src/web-extensions.c
diff --git a/src/meson.build b/src/meson.build
index f5f53e5..fda9f17 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -104,6 +104,7 @@ xdg_desktop_portal_sources = files(
   'settings.c',
   'trash.c',
   'wallpaper.c',
+  'web-extensions.c',
   'xdg-desktop-portal.c',
   'xdp-app-launch-context.c',
   'xdp-background-monitor.c',
diff --git a/src/web-extensions.c b/src/web-extensions.c
new file mode 100644
index 0000000..a33fee3
--- /dev/null
+++ b/src/web-extensions.c
@@ -0,0 +1,875 @@
+/*
+ * Copyright © 2022 Canonical Ltd
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "config.h"
+
+#include <stdint.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <glib/gi18n.h>
+#include <gio/gunixfdlist.h>
+#include <json-glib/json-glib.h>
+
+#include "gio/gdesktopappinfo.h"
+#include "gio/gio.h"
+#include "glib.h"
+#include "xdp-session.h"
+#include "web-extensions.h"
+#include "xdp-request.h"
+#include "xdp-permissions.h"
+#include "xdp-dbus.h"
+#include "xdp-impl-dbus.h"
+#include "xdp-utils.h"
+#include "xdp-app-info.h"
+
+#define PERMISSION_TABLE "webextensions"
+
+typedef struct _WebExtensions WebExtensions;
+typedef struct _WebExtensionsClass WebExtensionsClass;
+
+struct _WebExtensions
+{
+  XdpDbusWebExtensionsSkeleton parent_instance;
+};
+
+struct _WebExtensionsClass
+{
+  XdpDbusWebExtensionsSkeletonClass parent_class;
+};
+
+static XdpDbusImplAccess *access_impl;
+static WebExtensions *web_extensions;
+
+GType web_extensions_get_type (void);
+static void web_extensions_iface_init (XdpDbusWebExtensionsIface *iface);
+
+G_DEFINE_TYPE_WITH_CODE (WebExtensions, web_extensions, XDP_DBUS_TYPE_WEB_EXTENSIONS_SKELETON,
+                         G_IMPLEMENT_INTERFACE (XDP_DBUS_TYPE_WEB_EXTENSIONS,
+                                                web_extensions_iface_init));
+
+typedef enum _WebExtensionsSessionMode
+{
+  WEB_EXTENSIONS_SESSION_MODE_CHROMIUM,
+  WEB_EXTENSIONS_SESSION_MODE_MOZILLA,
+} WebExtensionsSessionMode;
+
+typedef enum _WebExtensionsSessionState
+{
+  WEB_EXTENSIONS_SESSION_STATE_INIT,
+  WEB_EXTENSIONS_SESSION_STATE_STARTING,
+  WEB_EXTENSIONS_SESSION_STATE_STARTED,
+  WEB_EXTENSIONS_SESSION_STATE_CLOSED,
+} WebExtensionsSessionState;
+
+typedef struct _WebExtensionsSession
+{
+  XdpSession parent;
+
+  WebExtensionsSessionMode mode;
+  WebExtensionsSessionState state;
+
+  GPid child_pid;
+  guint child_watch_id;
+
+  int standard_input;
+  int standard_output;
+  int standard_error;
+} WebExtensionsSession;
+
+typedef struct _WebExtensionsSessionClass
+{
+  XdpSessionClass parent_class;
+} WebExtensionsSessionClass;
+
+GType web_extensions_session_get_type (void);
+
+G_DEFINE_TYPE (WebExtensionsSession, web_extensions_session, xdp_session_get_type ());
+
+static void
+web_extensions_session_init (WebExtensionsSession *session)
+{
+  session->child_pid = -1;
+  session->child_watch_id = 0;
+
+  session->standard_input = -1;
+  session->standard_output = -1;
+  session->standard_error = -1;
+}
+
+static void
+web_extensions_session_close (XdpSession *session)
+{
+  WebExtensionsSession *web_extensions_session = (WebExtensionsSession *)session;
+
+  /* This function can be called repeatedly, e.g. by an explicit
+     "org.freedesktop.portal.Session::Close" message followed by
+     a call to finalize due to session's refcount reaching zero.
+  */
+  if (web_extensions_session->state == WEB_EXTENSIONS_SESSION_STATE_CLOSED) return;
+
+  /* We can assume that it is safe to transition to
+     WEB_EXTENSIONS_SESSION_STATE_CLOSED here, because we arrive
+     at web_extensions_session_close via one of two ways:
+
+     1. via the session_class->close function pointer from the
+        session_close function in src/session.c, which expects
+        its caller to lock the session's mutex (usually via the
+        SESSION_AUTOLOCK_UNREF macro); or
+     2. from web_extensions_session_finalize, at which point
+        the last reference to the session has been released.
+  */
+  web_extensions_session->state = WEB_EXTENSIONS_SESSION_STATE_CLOSED;
+
+  if (web_extensions_session->child_watch_id != 0)
+    {
+      g_source_remove (web_extensions_session->child_watch_id);
+      web_extensions_session->child_watch_id = 0;
+    }
+
+  if (web_extensions_session->child_pid > 0)
+    {
+      /* The responsibility of gracefully killing the process is
+         delegated to the browser, and the following SIGKILL is
+         a final attempt to clean up if necessary.
+      */
+      kill (web_extensions_session->child_pid, SIGKILL);
+      waitpid (web_extensions_session->child_pid, NULL, 0);
+      g_spawn_close_pid (web_extensions_session->child_pid);
+      web_extensions_session->child_pid = -1;
+    }
+
+  if (web_extensions_session->standard_input >= 0)
+    {
+      close (web_extensions_session->standard_input);
+      web_extensions_session->standard_input = -1;
+    }
+  if (web_extensions_session->standard_output >= 0)
+    {
+      close (web_extensions_session->standard_output);
+      web_extensions_session->standard_output = -1;
+    }
+  if (web_extensions_session->standard_error >= 0)
+    {
+      close (web_extensions_session->standard_error);
+      web_extensions_session->standard_error = -1;
+    }
+}
+
+static void
+web_extensions_session_finalize (GObject *object)
+{
+  XdpSession *session = (XdpSession *)object;
+
+  web_extensions_session_close (session);
+  G_OBJECT_CLASS (web_extensions_session_parent_class)->finalize (object);
+}
+
+static void
+web_extensions_session_class_init (WebExtensionsSessionClass *klass)
+{
+  /* Called at the first instantiation of WebExtensionsSession,
+     i.e. in web_extensions_session_new with the call to
+     g_initable_new.
+     https://docs.gtk.org/gobject/concepts.html#object-instantiation
+     https://docs.gtk.org/gio/type_func.Initable.new.html
+   */
+  GObjectClass *object_class;
+  XdpSessionClass *session_class;
+
+  object_class = G_OBJECT_CLASS (klass);
+  /* finalize is called when the session refcount reaches zero.
+     https://docs.gtk.org/gobject/concepts.html#reference-count
+  */
+  object_class->finalize = web_extensions_session_finalize;
+
+  session_class = (XdpSessionClass *)klass;
+  /* Register handler for org.freedesktop.portal.Session::Close */
+  session_class->close = web_extensions_session_close;
+}
+
+static WebExtensionsSession *
+web_extensions_session_new (GVariant *options,
+                            XdpCall *call,
+                            GDBusConnection *connection,
+                            GError **error)
+{
+  XdpSession *session;
+  WebExtensionsSession *web_extensions_session;
+  WebExtensionsSessionMode mode = WEB_EXTENSIONS_SESSION_MODE_MOZILLA;
+  const char *mode_str = NULL;
+  const char *session_token;
+
+  g_variant_lookup (options, "mode", "&s", &mode_str);
+  if (mode_str != NULL)
+    {
+      if (!strcmp(mode_str, "chromium"))
+        mode = WEB_EXTENSIONS_SESSION_MODE_CHROMIUM;
+      else if (!strcmp(mode_str, "mozilla"))
+        mode = WEB_EXTENSIONS_SESSION_MODE_MOZILLA;
+      else
+        {
+          g_set_error (error,
+                       XDG_DESKTOP_PORTAL_ERROR,
+                       XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
+                       "Invalid mode");
+          return NULL;
+        }
+    }
+
+  session_token = lookup_session_token (options);
+  session = g_initable_new (web_extensions_session_get_type (), NULL, error,
+                            "sender", call->sender,
+                            "app-id", xdp_app_info_get_id (call->app_info),
+                            "token", session_token,
+                            "connection", connection,
+                            NULL);
+
+  if (session)
+    g_debug ("webextensions session owned by '%s' created", session->sender);
+
+  if (!session)
+    {
+      g_warning ("Could not create WebExtensions session: %s", (*error)->message);
+      return NULL;
+    }
+  web_extensions_session = (WebExtensionsSession *)session;
+  web_extensions_session->mode = mode;
+  return web_extensions_session;
+}
+
+static gboolean
+handle_create_session (XdpDbusWebExtensions *object,
+                       GDBusMethodInvocation *invocation,
+                       GVariant *arg_options)
+{
+  XdpCall *call = xdp_call_from_invocation (invocation);
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  g_autoptr(GError) error = NULL;
+  XdpSession *session;
+
+  session = (XdpSession *)web_extensions_session_new (arg_options, call, connection, &error);
+  if (!session)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+  if (!xdp_session_export (session, &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      xdp_session_close (session, FALSE);
+      return TRUE;
+    }
+  xdp_session_register (session);
+
+  xdp_dbus_web_extensions_complete_create_session (object, invocation, session->id);
+
+  return TRUE;
+}
+
+static void
+on_host_exited (GPid pid,
+                gint status,
+                gpointer user_data)
+{
+  XdpSession *session = user_data;
+  WebExtensionsSession *web_extensions_session = (WebExtensionsSession *)session;
+
+  SESSION_AUTOLOCK (session);
+  web_extensions_session->child_pid = -1;
+  web_extensions_session->child_watch_id = 0;
+  xdp_session_close (session, TRUE);
+}
+
+static gboolean
+array_contains (JsonArray *array,
+                const char *value)
+{
+  guint length, i;
+
+  if (array == NULL)
+    return FALSE;
+
+  length = json_array_get_length (array);
+  for (i = 0; i < length; i++)
+    {
+      const char *element = json_array_get_string_element (array, i);
+      if (g_strcmp0 (element, value) == 0)
+        return TRUE;
+    }
+  return FALSE;
+}
+
+static gboolean
+is_valid_name (const char *name)
+{
+  /* This regexp comes from the Mozilla documentation on valid native
+     messaging host names:
+
+     https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_manifests#native_messaging_manifests
+
+     That is, one or more dot-separated groups composed of
+     alphanumeric characters and underscores.
+  */
+  return g_regex_match_simple ("^\\w+(\\.\\w+)*$", name, 0, 0);
+}
+
+static GStrv
+get_manifest_search_path (WebExtensionsSessionMode mode)
+{
+  /* IMPORTANT:
+     The safety model depends on the inability of the sandboxed
+     browser to write to the search locations specified below.
+
+     As this portal allows browser extensions to run a native
+     messaging application outside of the sandbox through the portal,
+     the sandboxing mechanism must ensure that these locations are
+     inaccessible to the browser.  If the locations are both readable
+     AND writable by the sandboxed browser, then a vulnerability
+     resulting in a file writing primitive within the sandbox could
+     result in arbitrary code execution outside of the sandbox through
+     the portal.
+
+     For example, the Firefox Snap package meets this criterion,
+     because all strictly confined Snap packages (including Firefox)
+     are prohibited by AppArmor from accessing most directories and
+     files in the user's home directory, except where explicitly
+     specified, for instance using the 'personal-files' interface.
+     https://snapcraft.io/docs/personal-files-interface
+  */
+  const char *hosts_path_str;
+  g_autoptr(GPtrArray) search_path = NULL;
+
+  hosts_path_str = g_getenv ("XDG_DESKTOP_PORTAL_WEB_EXTENSIONS_PATH");
+  if (hosts_path_str != NULL)
+    return g_strsplit (hosts_path_str, ":", -1);
+
+  search_path = g_ptr_array_new_with_free_func (g_free);
+  switch (mode)
+    {
+    case WEB_EXTENSIONS_SESSION_MODE_CHROMIUM:
+      /* Chrome and Chromium search paths documented here:
+       * https://developer.chrome.com/docs/extensions/nativeMessaging/#native-messaging-host-location
+       */
+      /* Add per-user directories */
+      g_ptr_array_add (search_path, g_build_filename (g_get_user_config_dir (), "google-chrome", "NativeMessagingHosts", NULL));
+      g_ptr_array_add (search_path, g_build_filename (g_get_user_config_dir (), "chromium", "NativeMessagingHosts", NULL));
+      /* Add system wide directories */
+      g_ptr_array_add (search_path, g_strdup ("/etc/opt/chrome/native-messaging-hosts"));
+      g_ptr_array_add (search_path, g_strdup ("/etc/chromium/native-messaging-hosts"));
+      /* And the same for xdg-desktop-portal's configured prefix */
+      g_ptr_array_add (search_path, g_strdup (SYSCONFDIR "/opt/chrome/native-messaging-hosts"));
+      g_ptr_array_add (search_path, g_strdup (SYSCONFDIR "/chromium/native-messaging-hosts"));
+      break;
+
+    case WEB_EXTENSIONS_SESSION_MODE_MOZILLA:
+      /* Firefox search paths documented here:
+       * https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_manifests#manifest_location
+       */
+      /* Add per-user directories */
+      g_ptr_array_add (search_path, g_build_filename (g_get_home_dir (), ".mozilla", "native-messaging-hosts", NULL));
+      g_ptr_array_add (search_path, g_build_filename (g_get_user_config_dir (), "mozilla", "native-messaging-hosts", NULL));
+      /* Add system wide directories */
+      g_ptr_array_add (search_path, g_strdup ("/usr/lib/mozilla/native-messaging-hosts"));
+      g_ptr_array_add (search_path, g_strdup ("/usr/lib64/mozilla/native-messaging-hosts"));
+      /* And the same for xdg-desktop-portal's configured prefix.
+         This is helpful on Debian-based systems where LIBDIR is
+         suffixed with 'dpkg-architecture -qDEB_HOST_MULTIARCH',
+         e.g. '/usr/lib/x86_64-linux-gnu'.
+         https://salsa.debian.org/debian/debhelper/-/blob/5b96b19b456fe5e094f2870327a753b4b3ece0dc/lib/Debian/Debhelper/Buildsystem/meson.pm#L78
+       */
+      g_ptr_array_add (search_path, g_strdup (LIBDIR "/mozilla/native-messaging-hosts"));
+      break;
+    }
+
+  g_ptr_array_add (search_path, NULL);
+  return (GStrv)g_ptr_array_free (g_steal_pointer (&search_path), FALSE);
+}
+
+static char *
+find_messaging_host (WebExtensionsSessionMode mode,
+                     const char *messaging_host_name,
+                     const char *extension_or_origin,
+                     char **out_description,
+                     char **out_manifest_filename,
+                     char **out_json_manifest,
+                     GError **error)
+{
+  g_auto(GStrv) search_path = NULL;
+  g_autoptr(JsonParser) parser = NULL;
+  g_autofree char *metadata_basename = NULL;
+  int i;
+
+  /* Check that the we have a valid native messaging host name */
+  if (!is_valid_name (messaging_host_name))
+    {
+      g_set_error (error,
+                   XDG_DESKTOP_PORTAL_ERROR,
+                   XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
+                   "Invalid native messaging host name");
+      return NULL;
+    }
+
+  search_path = get_manifest_search_path (mode);
+  parser = json_parser_new ();
+  metadata_basename = g_strconcat (messaging_host_name, ".json", NULL);
+
+  for (i = 0; search_path[i] != NULL; i++)
+    {
+      g_autofree char *metadata_filename = NULL;
+      g_autoptr(GError) load_error = NULL;
+      JsonObject *metadata_root;
+      const char *host_path;
+
+      metadata_filename = g_build_filename (search_path[i], metadata_basename, NULL);
+      if (!json_parser_load_from_file (parser, metadata_filename, &load_error))
+        {
+          /* If the file doesn't exist, continue searching. Error out
+             on anything else. */
+          if (g_error_matches (load_error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
+            continue;
+          g_propagate_error (error, g_steal_pointer (&load_error));
+          return NULL;
+        }
+
+      metadata_root = json_node_get_object (json_parser_get_root (parser));
+
+      /* Skip if metadata contains an unexpected name */
+      if (g_strcmp0 (json_object_get_string_member (metadata_root, "name"), messaging_host_name) != 0)
+        continue;
+
+      /* Skip if this is not a "stdio" type native messaging host */
+      if (g_strcmp0 (json_object_get_string_member (metadata_root, "type"), "stdio") != 0)
+        continue;
+
+      /* Skip if this host isn't available to the extension. Note
+       * that this ID is provided by the sandboxed browser, so this
+       * check is just to help implement its security policy. */
+      switch (mode)
+        {
+        case WEB_EXTENSIONS_SESSION_MODE_CHROMIUM:
+          if (!array_contains (json_object_get_array_member (metadata_root, "allowed_origins"), extension_or_origin))
+            continue;
+          break;
+        case WEB_EXTENSIONS_SESSION_MODE_MOZILLA:
+          if (!array_contains (json_object_get_array_member (metadata_root, "allowed_extensions"), extension_or_origin))
+            continue;
+          break;
+        }
+
+      host_path = json_object_get_string_member (metadata_root, "path");
+      if (!g_path_is_absolute (host_path))
+        {
+          g_set_error (error,
+                       XDG_DESKTOP_PORTAL_ERROR,
+                       XDG_DESKTOP_PORTAL_ERROR_FAILED,
+                       "Native messaging host path is not absolute");
+          return NULL;
+        }
+
+      /* Host matches: return its executable path and description */
+      if (out_description != NULL)
+        *out_description = g_strdup (json_object_get_string_member (metadata_root, "description"));
+      if (out_manifest_filename != NULL)
+        *out_manifest_filename = g_steal_pointer (&metadata_filename);
+      if (out_json_manifest != NULL)
+        *out_json_manifest = json_to_string (json_parser_get_root (parser), FALSE);
+      return g_strdup (host_path);
+    }
+
+  g_set_error (error,
+               XDG_DESKTOP_PORTAL_ERROR,
+               XDG_DESKTOP_PORTAL_ERROR_NOT_FOUND,
+               "Could not find native messaging host");
+  return NULL;
+}
+
+static gboolean
+handle_get_manifest (XdpDbusWebExtensions *object,
+                     GDBusMethodInvocation *invocation,
+                     const char *arg_session_handle,
+                     const char *arg_name,
+                     const char *arg_extension_or_origin)
+{
+  XdpCall *call = xdp_call_from_invocation (invocation);
+  XdpSession *session;
+  WebExtensionsSession *web_extensions_session;
+  g_autofree char *host_path = NULL;
+  g_autofree char *json_manifest = NULL;
+  g_autoptr(GError) error = NULL;
+
+  session = xdp_session_from_call (arg_session_handle, call);
+  if (!session)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_ACCESS_DENIED,
+                                             "Invalid session");
+      return TRUE;
+    }
+
+  SESSION_AUTOLOCK_UNREF (session);
+  web_extensions_session = (WebExtensionsSession *)session;
+
+  if (web_extensions_session->state != WEB_EXTENSIONS_SESSION_STATE_INIT)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_FAILED,
+                                             "Session already started");
+      return TRUE;
+    }
+
+  host_path = find_messaging_host (web_extensions_session->mode,
+                                   arg_name, arg_extension_or_origin,
+                                   NULL, NULL, &json_manifest, &error);
+  if (!host_path)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  xdp_dbus_web_extensions_complete_get_manifest (object, invocation, json_manifest);
+  return TRUE;
+}
+
+static void
+handle_start_in_thread (GTask *task,
+                        gpointer source_object,
+                        gpointer task_data,
+                        GCancellable *cancellable)
+{
+  XdpRequest *request = (XdpRequest *)task_data;
+  XdpSession *session;
+  WebExtensionsSession *web_extensions_session;
+  const char *arg_name;
+  char *arg_extension_or_origin;
+  const char *app_id;
+  g_autofree char *host_path = NULL;
+  g_autofree char *description = NULL;
+  g_autofree char *manifest_filename = NULL;
+  guint response = XDG_DESKTOP_PORTAL_RESPONSE_OTHER;
+  gboolean should_close_session;
+  XdpPermission permission;
+  gboolean allowed;
+  char *argv[] = {NULL, NULL, NULL, NULL};
+  g_autoptr(GError) error = NULL;
+
+  REQUEST_AUTOLOCK (request);
+  session = g_object_get_data (G_OBJECT (request), "session");
+  SESSION_AUTOLOCK_UNREF (g_object_ref (session));
+  g_object_set_data (G_OBJECT (request), "session", NULL);
+  web_extensions_session = (WebExtensionsSession *)session;
+
+  if (!request->exported || web_extensions_session->state != WEB_EXTENSIONS_SESSION_STATE_STARTING)
+    goto out;
+
+  arg_name = g_object_get_data (G_OBJECT (request), "name");
+  arg_extension_or_origin = g_object_get_data (G_OBJECT (request), "extension-or-origin");
+
+  host_path = find_messaging_host (web_extensions_session->mode,
+                                   arg_name, arg_extension_or_origin,
+                                   &description, &manifest_filename, NULL,
+                                   &error);
+  if (host_path == NULL)
+    {
+      g_warning ("Could not find WebExtensions backend: %s", error->message);
+      fflush(stderr);
+      fflush(stdout);
+      goto out;
+    }
+
+  app_id = xdp_app_info_get_id (request->app_info);
+  permission = xdp_get_permission_sync (app_id, PERMISSION_TABLE, arg_name);
+  if (permission == XDP_PERMISSION_ASK || permission == XDP_PERMISSION_UNSET)
+    {
+      guint access_response = 2;
+      g_autoptr(GVariant) access_results = NULL;
+      GVariantBuilder opt_builder;
+      const char *display_name;
+      g_autofree gchar *app_info_id = NULL;
+      g_autofree gchar *title = NULL;
+      g_autofree gchar *subtitle = NULL;
+      g_autofree gchar *body = NULL;
+      GAppInfo* info = xdp_app_info_get_gappinfo (request->app_info);
+
+      if (info)
+        {
+          g_auto(GStrv) app_id_components = g_strsplit (g_app_info_get_id (info), ".desktop", 2);
+          app_info_id = g_strdup (app_id_components[0]);
+        }
+      display_name = info ? g_app_info_get_display_name (info) : app_id;
+      title = g_strdup_printf (_("Allow %s to start WebExtension backend?"), display_name);
+      subtitle = g_strdup_printf (_("%s is requesting to launch \"%s\" (%s)."), display_name, description, arg_name);
+      body = g_strdup (_("This permission can be changed at any time from the privacy settings."));
+
+      g_variant_builder_init (&opt_builder, G_VARIANT_TYPE_VARDICT);
+      g_variant_builder_add (&opt_builder, "{sv}", "deny_label", g_variant_new_string (_("Don't allow")));
+      g_variant_builder_add (&opt_builder, "{sv}", "grant_label", g_variant_new_string (_("Allow")));
+      if (!xdp_dbus_impl_access_call_access_dialog_sync (access_impl,
+                                                         request->id,
+                                                         app_info_id ? app_info_id : app_id,
+                                                         "",
+                                                         title,
+                                                         subtitle,
+                                                         body,
+                                                         g_variant_builder_end (&opt_builder),
+                                                         &access_response,
+                                                         &access_results,
+                                                         cancellable,
+                                                         &error))
+        {
+          g_warning ("AccessDialog call failed: %s", error->message);
+          g_clear_error (&error);
+        }
+      allowed = access_response == 0;
+
+      // access_response == 2 means dialog has been closed
+      if (permission == XDP_PERMISSION_UNSET && access_response != 2)
+        xdp_set_permission_sync (app_id, PERMISSION_TABLE, arg_name, allowed ? XDP_PERMISSION_YES : XDP_PERMISSION_NO);
+    }
+  else
+    {
+      allowed = permission == XDP_PERMISSION_YES ? TRUE : FALSE;
+    }
+
+  if (!allowed)
+    {
+      response = XDG_DESKTOP_PORTAL_RESPONSE_CANCELLED;
+      goto out;
+    }
+
+  argv[0] = host_path;
+  switch (web_extensions_session->mode)
+    {
+    case WEB_EXTENSIONS_SESSION_MODE_CHROMIUM:
+      /* Pass the origin
+         https://developer.chrome.com/docs/extensions/develop/concepts/native-messaging
+       */
+      argv[1] = arg_extension_or_origin;
+      break;
+    case WEB_EXTENSIONS_SESSION_MODE_MOZILLA:
+      /* Pass the manifest filename and extension ID
+         https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_messaging
+         https://searchfox.org/mozilla-central/rev/9fcc11127fbfbdc88cbf37489dac90542e141c77/toolkit/components/extensions/NativeMessaging.sys.mjs#104-110
+       */
+      argv[1] = manifest_filename;
+      argv[2] = arg_extension_or_origin;
+      break;
+    }
+  if (!g_spawn_async_with_pipes (NULL, /* working_directory */
+                                 argv,
+                                 NULL, /* envp */
+                                 G_SPAWN_DO_NOT_REAP_CHILD,
+                                 NULL, /* child_setup */
+                                 NULL, /* user_data */
+                                 &web_extensions_session->child_pid,
+                                 &web_extensions_session->standard_input,
+                                 &web_extensions_session->standard_output,
+                                 &web_extensions_session->standard_error,
+                                 &error))
+    {
+      web_extensions_session->child_pid = -1;
+      goto out;
+    }
+
+  web_extensions_session->child_watch_id = g_child_watch_add_full (G_PRIORITY_DEFAULT,
+                                                                   web_extensions_session->child_pid,
+                                                                   on_host_exited,
+                                                                   g_object_ref (web_extensions_session),
+                                                                   g_object_unref);
+  web_extensions_session->state = WEB_EXTENSIONS_SESSION_STATE_STARTED;
+
+  response = XDG_DESKTOP_PORTAL_RESPONSE_SUCCESS;
+
+out:
+  should_close_session = !request->exported || response != XDG_DESKTOP_PORTAL_RESPONSE_SUCCESS;
+
+  if (request->exported)
+    {
+      GVariantBuilder results;
+
+      g_variant_builder_init (&results, G_VARIANT_TYPE_VARDICT);
+      xdp_dbus_request_emit_response (XDP_DBUS_REQUEST (request), response, g_variant_builder_end (&results));
+      xdp_request_unexport (request);
+    }
+
+  if (should_close_session)
+    xdp_session_close (session, TRUE);
+}
+
+static gboolean
+handle_start (XdpDbusWebExtensions *object,
+              GDBusMethodInvocation *invocation,
+              const char *arg_session_handle,
+              const char *arg_name,
+              const char *arg_extension_or_origin,
+              GVariant *arg_options)
+{
+  XdpRequest *request = xdp_request_from_invocation (invocation);
+  XdpSession *session;
+  WebExtensionsSession *web_extensions_session;
+  g_autoptr(GTask) task = NULL;
+
+  REQUEST_AUTOLOCK (request);
+
+  session = xdp_session_from_request (arg_session_handle, request);
+  if (!session)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_ACCESS_DENIED,
+                                             "Invalid session");
+      return TRUE;
+    }
+
+  SESSION_AUTOLOCK_UNREF (session);
+  web_extensions_session = (WebExtensionsSession *)session;
+
+  if (web_extensions_session->state != WEB_EXTENSIONS_SESSION_STATE_INIT)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_FAILED,
+                                             "Session already started");
+      return TRUE;
+    }
+
+  web_extensions_session->state = WEB_EXTENSIONS_SESSION_STATE_STARTING;
+  g_object_set_data_full (G_OBJECT (request), "session", g_object_ref (session), g_object_unref);
+  g_object_set_data_full (G_OBJECT (request), "name", g_strdup (arg_name), g_free);
+  g_object_set_data_full (G_OBJECT (request), "extension-or-origin", g_strdup (arg_extension_or_origin), g_free);
+
+  xdp_request_export (request, g_dbus_method_invocation_get_connection (invocation));
+  xdp_dbus_web_extensions_complete_start (object, invocation, request->id);
+
+  task = g_task_new (object, NULL, NULL, NULL);
+  g_task_set_task_data (task, g_object_ref (request), g_object_unref);
+  g_task_run_in_thread (task, handle_start_in_thread);
+
+  return TRUE;
+}
+
+
+static gboolean
+handle_get_pipes (XdpDbusWebExtensions *object,
+                  GDBusMethodInvocation *invocation,
+                  GUnixFDList *fd_list,
+                  const char *arg_session_handle,
+                  GVariant *arg_options)
+{
+  XdpCall *call = xdp_call_from_invocation (invocation);
+  XdpSession *session;
+  WebExtensionsSession *web_extensions_session;
+  int fds[3];
+  g_autoptr(GUnixFDList) out_fd_list = NULL;
+
+  session = xdp_session_from_call (arg_session_handle, call);
+  if (!session)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_ACCESS_DENIED,
+                                             "Invalid session");
+      return TRUE;
+    }
+
+  SESSION_AUTOLOCK_UNREF (session);
+  web_extensions_session = (WebExtensionsSession *)session;
+
+  if (web_extensions_session->state != WEB_EXTENSIONS_SESSION_STATE_STARTED)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_FAILED,
+                                             "Session not started");
+      return TRUE;
+    }
+
+  if (web_extensions_session->standard_input < 0 ||
+      web_extensions_session->standard_output < 0 ||
+      web_extensions_session->standard_error < 0)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_FAILED,
+                                             "GetPipes already called");
+      return TRUE;
+    }
+
+  fds[0] = web_extensions_session->standard_input;
+  fds[1] = web_extensions_session->standard_output;
+  fds[2] = web_extensions_session->standard_error;
+  out_fd_list = g_unix_fd_list_new_from_array (fds, G_N_ELEMENTS (fds));
+  /* out_fd_list now owns the file descriptors */
+  web_extensions_session->standard_input = -1;
+  web_extensions_session->standard_output = -1;
+  web_extensions_session->standard_error = -1;
+
+  xdp_dbus_web_extensions_complete_get_pipes (object, invocation, out_fd_list,
+                                              g_variant_new_handle (0),
+                                              g_variant_new_handle (1),
+                                              g_variant_new_handle (2));
+  return TRUE;
+}
+
+static void
+web_extensions_iface_init (XdpDbusWebExtensionsIface *iface)
+{
+  iface->handle_create_session = handle_create_session;
+  iface->handle_get_manifest = handle_get_manifest;
+  iface->handle_start = handle_start;
+  iface->handle_get_pipes = handle_get_pipes;
+}
+
+static void
+web_extensions_init (WebExtensions *web_extensions)
+{
+  xdp_dbus_web_extensions_set_version (XDP_DBUS_WEB_EXTENSIONS (web_extensions), 1);
+}
+
+static void
+web_extensions_class_init (WebExtensionsClass *klass)
+{
+  /* Called at the first instantiation of WebExtensions,
+     i.e. in web_extensions_create with the call to g_object_new.
+     https://docs.gtk.org/gobject/concepts.html#object-instantiation
+   */
+}
+
+GDBusInterfaceSkeleton *
+web_extensions_create (GDBusConnection *connection,
+                       const char *dbus_name_access)
+{
+  g_autoptr(GError) error = NULL;
+
+  web_extensions = g_object_new (web_extensions_get_type (), NULL);
+
+  access_impl = xdp_dbus_impl_access_proxy_new_sync (connection,
+                                                     G_DBUS_PROXY_FLAGS_NONE,
+                                                     dbus_name_access,
+                                                     DESKTOP_PORTAL_OBJECT_PATH,
+                                                     NULL,
+                                                     &error);
+  if (access_impl == NULL)
+    {
+      g_warning ("Failed to create access proxy: %s", error->message);
+      return NULL;
+    }
+
+  return G_DBUS_INTERFACE_SKELETON (web_extensions);
+}
diff --git a/src/web-extensions.h b/src/web-extensions.h
new file mode 100644
index 0000000..72b947a
--- /dev/null
+++ b/src/web-extensions.h
@@ -0,0 +1,24 @@
+/*
+ * Copyright © 2022 Canonical Ltd
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#pragma once
+
+#include <gio/gio.h>
+
+GDBusInterfaceSkeleton *web_extensions_create (GDBusConnection *connection,
+                                               const char *access_dbus_name);
diff --git a/src/xdg-desktop-portal.c b/src/xdg-desktop-portal.c
index edaebe1..a602d3c 100644
--- a/src/xdg-desktop-portal.c
+++ b/src/xdg-desktop-portal.c
@@ -70,6 +70,7 @@
 #include "trash.h"
 #include "usb.h"
 #include "wallpaper.h"
+#include "web-extensions.h"
 
 static int global_exit_status = 0;
 static GMainLoop *loop = NULL;
@@ -315,6 +316,9 @@ on_bus_acquired (GDBusConnection *connection,
     {
       XdpPortalImplementation *tmp;
 
+      export_portal_implementation (connection,
+                                    web_extensions_create (connection,
+                                                           access_impl->dbus_name));
 #ifdef HAVE_GEOCLUE
       export_portal_implementation (connection,
                                     location_create (connection,
diff --git a/src/xdp-request.c b/src/xdp-request.c
index 6aaa7ed..46ceca8 100644
--- a/src/xdp-request.c
+++ b/src/xdp-request.c
@@ -188,6 +188,18 @@ get_token (GDBusMethodInvocation *invocation)
       if (method_info->option_arg >= 0)
         options = g_variant_get_child_value (parameters, method_info->option_arg);
     }
+  else if (strcmp (interface, "org.freedesktop.portal.WebExtensions") == 0)
+    {
+      if (strcmp (method, "Start") == 0)
+        {
+          options = g_variant_get_child_value (parameters, 3);
+        }
+      else
+        {
+          g_warning ("Support for %s::%s missing in %s",
+                     interface, method, G_STRLOC);
+        }
+    }
   else
     {
       g_warning ("Support for %s::%s missing in %s",
diff --git a/tests/conftest.py b/tests/conftest.py
index 1163e5a..6241411 100644
--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -120,16 +120,31 @@ def xdg_data_home_files() -> Dict[str, bytes]:
     return {}
 
 
+@pytest.fixture
+def xdg_config_home_files() -> Dict[str, bytes]:
+    """
+    Default fixture which can be used to create files in the temporary
+    XDG_CONFIG_HOME directory of the test.
+    """
+    return {}
+
+
 @pytest.fixture(autouse=True)
-def ensure_xdg_data_home(
-    create_test_dirs: Any, xdg_data_home_files: Dict[str, bytes]
+def ensure_xdg_test_dirs_files(
+    create_test_dirs: Any,
+    xdg_data_home_files: Dict[str, bytes],
+    xdg_config_home_files: Dict[str, bytes],
 ) -> None:
-    files = xdg_data_home_files
-    for name, content in files.items():
-        file_path = Path(os.environ["XDG_DATA_HOME"]) / name
-        file_path.parent.mkdir(parents=True, exist_ok=True)
-        with open(file_path.absolute().as_posix(), "wb") as f:
-            f.write(content)
+    def ensure_files(files, xdg_test_dir):
+        dir_path = Path(os.environ[xdg_test_dir])
+        for name, content in files.items():
+            file_path = dir_path / name
+            file_path.parent.mkdir(parents=True, exist_ok=True)
+            with open(file_path.absolute().as_posix(), "wb") as f:
+                f.write(content)
+
+    ensure_files(xdg_data_home_files, "XDG_DATA_HOME")
+    ensure_files(xdg_config_home_files, "XDG_CONFIG_HOME")
 
 
 @pytest.fixture
diff --git a/tests/meson.build b/tests/meson.build
index 87b2e39..23b7c1f 100644
--- a/tests/meson.build
+++ b/tests/meson.build
@@ -105,6 +105,7 @@ pytest_files = [
   'test_trash.py',
   'test_usb.py',
   'test_wallpaper.py',
+  'test_webextensions.py',
 ]
 
 template_files = [
diff --git a/tests/test_webextensions.py b/tests/test_webextensions.py
new file mode 100644
index 0000000..fc6b682
--- /dev/null
+++ b/tests/test_webextensions.py
@@ -0,0 +1,147 @@
+import tests as xdp
+import random
+import dbus
+import json
+import pytest
+
+
+MANIFEST = b"""
+{
+  "name": "org.example.testing",
+  "description": "Test native messaging host",
+  "path": "/usr/bin/cat",
+  "type": "stdio",
+  "allowed_extensions": [
+    "some-extension@example.org"
+  ]
+}
+"""
+
+
+@pytest.fixture
+def xdg_config_home_files():
+    return {
+        "mozilla/native-messaging-hosts/org.example.testing.json": MANIFEST,
+    }
+
+
+def set_web_extensions_permissions(permission, name, app_id, dbus_con):
+    permission_store = xdp.get_permission_store_iface(dbus_con)
+    try:
+        permission_store.SetPermission(
+            "webextensions", True, "org.example.testing", app_id, [permission]
+        )
+    except dbus.DBusException as e:
+        raise RuntimeError(f"Failed to set permissions: {e}")
+
+
+class TestWebExtensions:
+    def test_version(self, portals, dbus_con):
+        xdp.check_version(dbus_con, "WebExtensions", 1)
+
+    def create_session(self, dbus_con):
+        webextensions_intf = xdp.get_portal_iface(dbus_con, "WebExtensions")
+        session_token = "portal" + str(random.randint(0, 100000))
+        options = {"mode": "mozilla", "session_handle_token": session_token}
+        session = xdp.Session(
+            dbus_con,
+            webextensions_intf.CreateSession(options),
+        )
+        return session.handle
+
+    def get_manifest(self, name, extension_or_origin, dbus_con):
+        webextensions_intf = xdp.get_portal_iface(dbus_con, "WebExtensions")
+        return webextensions_intf.GetManifest(
+            self.session_handle, name, extension_or_origin
+        )
+
+    def start(self, name, extension_or_origin, dbus_con):
+        webextensions_intf = xdp.get_portal_iface(dbus_con, "WebExtensions")
+        request = xdp.Request(dbus_con, webextensions_intf)
+        return request.call(
+            "Start",
+            session_handle=self.session_handle,
+            name=name,
+            extension_or_origin=extension_or_origin,
+            options={},
+        )
+
+    def get_pipes(self, dbus_con):
+        webextensions_intf = xdp.get_portal_iface(dbus_con, "WebExtensions")
+        response = webextensions_intf.GetPipes(self.session_handle, {})
+        return response
+
+    def test_web_extensions_basic(self, portals, app_id, dbus_con):
+        name = "org.example.testing"
+        set_web_extensions_permissions("yes", name, app_id, dbus_con)
+
+        extension_or_origin = "some-extension@example.org"
+
+        self.session_handle = self.create_session(dbus_con)
+        assert self.session_handle, "Session handle should not be None"
+
+        manifest_data = self.get_manifest(name, extension_or_origin, dbus_con)
+        assert manifest_data, "Manifest should not be None"
+
+        manifest = json.loads(manifest_data)
+        assert (
+            "some-extension@example.org" in manifest["allowed_extensions"]
+        ), "Extension ID missing in manifest"
+
+        response = self.start(name, extension_or_origin, dbus_con)
+        assert response, "Response cannot be None"
+        assert response.response == 0, "Returned invalid response code, expected 0"
+
+        pipes = self.get_pipes(dbus_con)
+        assert pipes, "No pipes returned"
+        assert pipes[0].take() > 0, "stdin file descriptor should be valid"
+        assert pipes[1].take() > 0, "stdout file descriptor should be valid"
+        assert pipes[2].take() > 0, "stderr file descriptor should be valid"
+
+    def test_web_extensions_denied(self, portals, app_id, dbus_con):
+        name = "org.example.testing"
+        set_web_extensions_permissions("no", name, app_id, dbus_con)
+
+        extension_or_origin = "some-extension@example.org"
+
+        self.session_handle = self.create_session(dbus_con)
+        assert self.session_handle, "Session handle should not be None"
+
+        manifest_data = self.get_manifest(name, extension_or_origin, dbus_con)
+        assert manifest_data, "Manifest should not be None"
+
+        manifest = json.loads(manifest_data)
+        assert (
+            "some-extension@example.org" in manifest["allowed_extensions"]
+        ), "Extension ID missing in manifest"
+
+        response = self.start(name, extension_or_origin, dbus_con)
+        print(response)
+        assert response, "Response cannot be None"
+        assert response.response == 1, "Returned invalid response code, expected 1"
+
+        pipes = None
+        try:
+            pipes = self.get_pipes(dbus_con)
+        except dbus.DBusException as e:
+            assert (
+                e.get_dbus_name() == "org.freedesktop.DBus.Error.AccessDenied"
+            ), "unknown exception"
+        assert pipes is None, "Expected pipes is None"
+
+    def test_web_extensions_bad_name(self, portals, app_id, dbus_con):
+        bad_extension_names = ["no-dashes", "../foo", "no_trailing_dot."]
+
+        name = "org.example.testing"
+        set_web_extensions_permissions("yes", name, app_id, dbus_con)
+
+        self.session_handle = self.create_session(dbus_con)
+        assert self.session_handle, "Session handle should not be None"
+
+        extension_or_origin = "some-extension@example.org"
+
+        for extension_or_origin in bad_extension_names:
+            self.session_handle = self.create_session(dbus_con)
+            response = self.start(name, extension_or_origin, dbus_con)
+            assert response, "Response cannot be None"
+            assert response.response == 2, "Returned invalid response, expected 2"
